File: src/HookBridge.Dashboard/src/types/subscription.ts
import type { PagedRequest } from './pagination';

export type BackoffType = 'Fixed' | 'Exponential';

export type AuthenticationType = 'None' | 'Basic' | 'ApiKeyHeader' | 'HmacSignature' | 'OAuth2ClientCredentials';

export type KeyValue = {
  name: string;
  value: string;
};

export type RetryPolicy = {
  maxAttempts: number;
  initialDelaySeconds: number;
  backoffType: BackoffType;
};

export type BasicAuthentication = {
  username: string;
  password: string;
};

export type ApiKeyHeaderAuthentication = {
  headerName: string;
  headerValue: string;
};

export type HmacSignatureAuthentication = {
  secret: string;
  headerName: string;
  algorithm: string;
};

export type OAuth2ClientCredentialsAuthentication = {
  tokenUrl: string;
  clientId: string;
  clientSecret: string;
  scope?: string;
};

export type Authentication = {
  type: AuthenticationType;
  basic?: BasicAuthentication;
  apiKeyHeader?: ApiKeyHeaderAuthentication;
  hmacSignature?: HmacSignatureAuthentication;
  oauth2?: OAuth2ClientCredentialsAuthentication;
};

export type Subscription = {
  id: string;
  eventType: string;
  targetUrl: string;
  headers: KeyValue[];
  authentication?: Authentication;
  retryPolicy: RetryPolicy;
  timeoutSeconds: number;
  deliveryFormat: 'Raw' | 'HookBridgeEnvelope' | 'CloudEventsStructured' | 'CloudEventsBinary';
  isActive: boolean;
  disabledAt?: string;
  createdAt: string;
  updatedAt?: string;
};

export type SubscriptionListFilters = PagedRequest & {
  eventType?: string;
  targetUrl?: string;
  isActive?: boolean;
};

export type CreateSubscriptionRequest = {
  eventType?: string;
  deliveryFormat?: 'Raw' | 'HookBridgeEnvelope' | 'CloudEventsStructured' | 'CloudEventsBinary';
  targetUrl: string;
  headers: KeyValue[];
  authentication?: Authentication;
  retryPolicy: RetryPolicy;
  timeoutSeconds: number;
};

export type UpdateSubscriptionRequest = {
  eventType?: string;
  deliveryFormat?: 'Raw' | 'HookBridgeEnvelope' | 'CloudEventsStructured' | 'CloudEventsBinary';
  targetUrl: string;
  headers: KeyValue[];
  authentication?: Authentication;
  retryPolicy: RetryPolicy;
  timeoutSeconds: number;
};

export type EndpointValidationRequest = {
  targetUrl: string;
  method?: string;
  samplePayload?: Record<string, unknown>;
  headers?: KeyValue[];
  authentication?: Authentication;
  timeoutSeconds: number;
};

export type EndpointValidationResponse = {
  isSuccess: boolean;
  statusCode?: number;
  message: string;
  durationMs: number;
  responseBody?: string;
};

File: src/HookBridge.Domain/Entities/Subscription.cs
namespace HookBridge.Domain.Entities;

/// <summary>
/// Represents a tenant subscription that receives specific event types via webhook.
/// </summary>
public sealed class Subscription : BaseEntity
{
    public string TenantId { get; set; } = string.Empty;

    public string EventType { get; set; } = string.Empty;

    public string TargetUrl { get; set; } = string.Empty;

    public List<KeyValueItem> Headers { get; set; } = [];

    public AuthenticationConfig? Authentication { get; set; }

    public RetryPolicy RetryPolicy { get; set; } = new();

    public int TimeoutSeconds { get; set; }

    public bool IsActive { get; set; }

    public DateTime? DisabledAt { get; set; }
}

public sealed class KeyValueItem
{
    public string Name { get; set; } = string.Empty;

    public string Value { get; set; } = string.Empty;
}

public sealed class RetryPolicy
{
    public int MaxAttempts { get; set; }

    public int InitialDelaySeconds { get; set; }

    public string BackoffType { get; set; } = string.Empty;
}

public sealed class AuthenticationConfig
{
    public string Type { get; set; } = string.Empty;

    public BasicAuthConfig? Basic { get; set; }

    public OAuth2ClientCredentialsConfig? OAuth2 { get; set; }

    public ApiKeyHeaderConfig? ApiKeyHeader { get; set; }

    public HmacSignatureConfig? HmacSignature { get; set; }
}

public sealed class BasicAuthConfig
{
    public string Username { get; set; } = string.Empty;

    public string Password { get; set; } = string.Empty;
}

public sealed class OAuth2ClientCredentialsConfig
{
    public string TokenUrl { get; set; } = string.Empty;

    public string ClientId { get; set; } = string.Empty;

    public string ClientSecret { get; set; } = string.Empty;

    public string? Scope { get; set; }
}

public sealed class ApiKeyHeaderConfig
{
    public string HeaderName { get; set; } = string.Empty;

    public string HeaderValue { get; set; } = string.Empty;
}

public sealed class HmacSignatureConfig
{
    public string Secret { get; set; } = string.Empty;

    public string HeaderName { get; set; } = string.Empty;

    public string Algorithm { get; set; } = string.Empty;
}
